### PR TITLE
[SPARK-29454][SQL] Reduce unsafeProjection times when read Parquet file use non-vectorized mode

### PR DESCRIPTION
**Pull request by LuciferYang**
### What changes were proposed in this pull request?

There will be 2 times unsafeProjection convert operation When we read a Parquet data file use non-vectorized mode:

1.  `ParquetGroupConverter` call unsafeProjection function to covert `SpecificInternalRow` to `UnsafeRow` every times when read Parquet data file use `ParquetRecordReader`.

2. `ParquetFileFormat` will call unsafeProjection function to covert this `UnsafeRow` to another `UnsafeRow` again when partitionSchema is not empty in DataSourceV1 branch, and `PartitionReaderWithPartitionValues` will  always do this convert operation in DataSourceV2 branch.

In this pr,  remove `unsafeProjection` convert operation in `ParquetGroupConverter` and change `ParquetRecordReader`  to produce `SpecificInternalRow`  instead of `UnsafeRow`.

### Why are the changes needed?
The first time convert in `ParquetGroupConverter` is redundant and `ParquetRecordReader` return a `InternalRow(SpecificInternalRow)` is enough.

### Does this PR introduce any user-facing change?
No.

### How was this patch tested?

Unit Test

### CODE REVIEW

**Comment by LuciferYang** [comment]
@xuanyuanking Could you take a look at the PR, thx~

**Comment by dongjoon-hyun** [comment]
ok to test

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFileFormat.scala [hunk 1]
```diff
@@ -328,32 +328,28 @@ class ParquetFileFormat
         iter.asInstanceOf[Iterator[InternalRow]]
       } else {
         logDebug(s"Falling back to parquet-mr")
```
Comment (dongjoon-hyun) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFileFormat.scala]: So, is this PR aiming only for non-vectorized code, @LuciferYang ?

Comment (LuciferYang) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFileFormat.scala]: ok~

**Comment by cloud-fan** [comment]
LGTM, merging to master!

**Comment by LuciferYang** [comment]
thx @dongjoon-hyun @cloud-fan ~

**Pull request closed without merging**